clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -1407,6 +1423,14 @@ class SyclKernelDeclCreator : public SyclKernelFieldHandler {
     PointeeTy = SemaRef.getASTContext().getQualifiedType(
         PointeeTy.getUnqualifiedType(), Quals);
     QualType ModTy = SemaRef.getASTContext().getPointerType(PointeeTy);
+    // When the kernel is generated, struct type captures are decomposed; i.e.
+    // the parameters of the kernel are the fields of the struct, and not the
+    // struct itself. This causes an error in the backend when the struct field
+    // is a pointer, since non-USM pointers cannot be passed directly.
+    // To work around this issue, all pointers are wrapped in a generated
+    // 'wrapper_class'.
```

Comment:
They end up being different classes, lookup is already done at this point, so the names stop mattering, only their position in the AST.  I fear what the IR would look like however, since we've given them a name.

Reply:
I guess IR should be fine, I think it will just apply some numbers to the name like it does with the same names in kernel arguments, i.e wrapper_class42...

Reply:
I've added a test for this. Looks like there is no issue. LLVM IR handles it by naming classes wrapper_class, wrapper_class.0, etc